components/theia/packages/gitpod-extension/src/browser/gitpod-user-storage-cached.ts

```diff
@@ -0,0 +1,100 @@
+/**
+ * Copyright (c) 2020 TypeFox GmbH. All rights reserved.
+ * Licensed under the GNU Affero General Public License (AGPL).
+ * See License-AGPL.txt in the project root for license information.
+ */
+
+import { injectable, inject, postConstruct } from "inversify";
+import { GitpodServiceProvider } from "./gitpod-service-provider";
+import { GitpodServer } from "@gitpod/gitpod-protocol";
+import { Disposable } from "@theia/core";
+import { JsonRpcProxy } from "@gitpod/gitpod-protocol/lib/messaging/proxy-factory";
+
+/**
+ * To avoid having badly written extensions executing too much requests and putting pressure on other components we
+ * cache writes and reads to the GitpodUserStorage.
+ */
+@injectable()
+export class CachedUserStorage implements Disposable {
+
+    static INTERVAL_SECONDS = 10;
+
+    @inject(GitpodServiceProvider) protected serviceProvider: GitpodServiceProvider;
+
+    protected readonly contentCache = new Map<string, string>();
+    protected updatedSinceLastWrite = new Set<string>();
+
+    protected timout: NodeJS.Timeout | undefined;
+
+
+    @postConstruct()
+    initialize() {
+        this.startPeriodicCacheWriter(CachedUserStorage.INTERVAL_SECONDS);
+    }
+
+    async write(resource: GitpodServer.UpdateUserStorageResourceOptions): Promise<void> {
+        this.contentCache.set(resource.uri, resource.content);
+        this.updatedSinceLastWrite.add(resource.uri);
+        
+        console.count(`user storage write to: ${resource.uri}`);
+    }
+
+    async read(req: GitpodServer.GetUserStorageResourceOptions): Promise<string> {
+        const cached = this.contentCache.get(req.uri);
+        if (cached !== undefined) {
+            return cached;
+        }
+
+        const content = await this.getServer().getUserStorageResource(req);
+        this.contentCache.set(req.uri, content);
```

Comment:
Should not we check cache here again? in the meantime it could be updated?

Reply:
Yes. This is the very reason I wanted to avoid the "update on read" in the first place - above the others were in for it. I find without it better as well.

Reply:
Simply not writing if `this.contentCache.has(req.uri)` should be sufficient.